src/write/elf/object.rs

```diff
@@ -766,6 +768,27 @@ impl<'a> Object<'a> {
                                 return Err(Error(format!("unimplemented relocation {:?}", reloc)));
                             }
                         },
+                        Architecture::Sharc => match (reloc.kind, reloc.encoding, reloc.size) {
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs32TypeA, 48) => elf::R_SHARC_ADDR32_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs32Data, 32) => elf::R_SHARC_ADDR_VAR_V3,
+                            (RelocationKind::Relative, RelocationEncoding::SharcPcr24TypeA, 48) => elf::R_SHARC_PCRLONG_V3,
+                            (RelocationKind::Relative, RelocationEncoding::SharcPcr6TypeA, 48) => elf::R_SHARC_PCRSHORT_V3,
+                            (RelocationKind::Relative, RelocationEncoding::SharcPcr6TypeB, 32) => elf::R_SHARC_PCRSHORT_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs16Data, 16) => elf::R_SHARC_ADDR_VAR16_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs16TypeA, 48) => elf::R_SHARC_DATA16_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs16TypeB, 32) => elf::R_SHARC_DATA16_VISA_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs24TypeA, 48) => elf::R_SHARC_ADDR24_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs6TypeA, 48) => elf::R_SHARC_DATA6_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs6TypeB, 32) => elf::R_SHARC_DATA6_VISA_V3,
+                            (RelocationKind::Absolute, RelocationEncoding::SharcAbs7TypeB, 32) => elf::R_SHARC_DATA7_VISA_V3,
+                            (_, RelocationEncoding::Generic, _) => {
+                                return Err(Error(format!("SHARC+ ISA does not have a generic relocation encoding")));
```

Comment:
General rule for SHARC: If you find it surprising, you're probably reading it correctly.
It's a good question though, I haven't looked at any DWARF output yet; I don't need it and I don't know much about DWARF. I'll put it on my TODO list.

AFAICT from reversing, there is an R_SHARC_* constant for each of these permutations. I've used the official compiler to emit every instruction in the ISA which can accept a relocation, and these are the values it uses – there is nowhere for a generic option to go.

Reply:
I would guess that `R_SHARC_ADDR_VAR_V3` is used for DWARF. Since we are using `RelocationEncoding::Generic` for `R_SHARC_ADDR_VAR_V3` now, I don't think this error message makes sense anymore. It would be valid to use RelocationKind::Elf(`R_SHARC_ADDR_VAR_V3`) with `RelocationEncoding::Generic` and this match pattern prevents that, so I think this match pattern should be deleted.

Reply:
I still think that this match pattern for `RelocationEncoding::Generic` should be deleted.

Reply:
Yeah, I agree. I'll delete that now.